File: compiler/bench/rabinkarp.c
// expected: 0
// Rabin-Karp search, 100K searches
// Checksum: sum of found positions

#include <stdio.h>

char text[64];
char pattern[8];
int pat_len;
#define PRIME 101
#define BASE 256

void fill_text(int n) {
    for (int i = 0; i < 32; i++) {
        text[i] = 'a' + ((i + n) % 26);
    }
}

void fill_pat(int n) {
    // Pattern is 3 consecutive letters starting at offset n%26
    for (int i = 0; i < 3; i++) {
        pattern[i] = 'a' + ((n + i) % 26);
    }
    pat_len = 3;
}

int pat_hash(void) {
    int h = 0;
    for (int i = 0; i < pat_len; i++) {
        h = (h * BASE + pattern[i]) % PRIME;
    }
    return h;
}

int text_hash(int pos) {
    int h = 0;
    for (int i = 0; i < pat_len; i++) {
        h = (h * BASE + text[pos + i]) % PRIME;
    }
    return h;
}

int match_at(int pos) {
    for (int i = 0; i < pat_len; i++) {
        if (text[pos + i] != pattern[i]) return 0;
    }
    return 1;
}

int rk_search(void) {
    int ph = pat_hash();
    int th = text_hash(0);
    if (ph == th && match_at(0)) return 0;

    int h = 1;
    for (int i = 0; i < pat_len - 1; i++) h = (h * BASE) % PRIME;

    for (int i = 1; i <= 32 - pat_len; i++) {
        th = ((th - text[i-1] * h % PRIME + PRIME) * BASE + text[i + pat_len - 1]) % PRIME;
        if (th == ph && match_at(i)) return i;
    }
    return -1;
}

int main(void) {
    long sum = 0;
    for (int i = 0; i < 100000; i++) {
        fill_text(i);
        fill_pat(i);
        int pos = rk_search();
        if (pos >= 0) sum += pos;
    }
    printf("%ld\n", sum);
    return 0;
}

File: compiler/bench/rabinkarp.fs
\ expected: 0
\ Rabin-Karp search, 100K searches
\ Checksum: sum of found positions

create text 64 allot
create pattern 8 allot
variable pat-len
101 constant PRIME
256 constant BASE

: fill-text ( n -- ) 32 0 do i over + 26 mod 97 + text i + c! loop drop ;
: fill-pat ( n -- ) 3 0 do over i + 26 mod 97 + pattern i + c! loop drop 3 pat-len ! ;

: pat-hash ( -- h )
  0 pat-len @ 0 do BASE * pattern i + c@ + PRIME mod loop ;

: text-hash ( pos -- h )
  0 pat-len @ 0 do BASE * text 2 pick i + + c@ + PRIME mod loop swap drop ;

: match-at ( pos -- 0|1 )
  1 pat-len @ 0 do text over i + + c@ pattern i + c@ <> if drop 0 leave then loop swap drop ;

: rk-h ( -- h ) 1 pat-len @ 1- 0 do BASE * PRIME mod loop ;

: rk-search ( -- pos )
  pat-hash
  0 text-hash
  over over = if 0 match-at if 2drop 0 exit then then
  rk-h
  32 pat-len @ - 0 do
    text j + c@ over * PRIME mod -
    PRIME + PRIME mod
    BASE * text j pat-len @ + + c@ + PRIME mod
    2 pick over = if
      j 1+ match-at if
        2drop drop j 1+ unloop exit
      then
    then
  loop
  2drop drop -1 ;

: bench-rk ( -- sum )
  0
  100000 0 do
    i fill-text
    i fill-pat
    rk-search
    dup 0< if drop else + then
  loop ;

: main bench-rk . cr ;
